subsys/usb/device_next/class/usbd_uac2.c

```diff
@@ -0,0 +1,799 @@
+				      sys_le16_to_cpu(data_ep->wMaxPacketSize),
+				      cfg->as_terminals[as_idx]);
+
+		fb_ep = get_as_feedback_ep(node, as_idx);
+		if (fb_ep) {
+			write_explicit_feedback(node, fb_ep->bEndpointAddress,
+						cfg->as_terminals[as_idx]);
+		}
+	}
+}
+
+/* Table 5-6: 4-byte Control CUR Parameter Block */
+static void layout3_cur_response(struct net_buf *const buf, uint16_t length,
+				 const uint32_t value)
+{
+	uint8_t tmp[4];
+
+	/* dCUR */
+	sys_put_le32(value, tmp);
+	net_buf_add_mem(buf, tmp, MIN(length, 4));
+}
+
+/* Table 5-7: 4-byte Control RANGE Parameter Block */
+static void layout3_range_response(struct net_buf *const buf, uint16_t length,
+				   const uint32_t *min, const uint32_t *max,
+				   const uint32_t *res, int n)
+{
+	uint16_t to_add;
+	uint8_t tmp[4];
+	int i;
+	int item;
+
+	/* wNumSubRanges */
+	sys_put_le16(n, tmp);
+	to_add = MIN(length, 2);
+	net_buf_add_mem(buf, tmp, to_add);
+	length -= to_add;
+
+	/* Keep adding dMIN, dMAX, dRES as long as we have entries to add and
+	 * we didn't reach wLength response limit.
+	 */
+	i = item = 0;
+	while ((length > 0) && (i < n)) {
+		to_add = MIN(length, 4);
+		if (item == 0) {
+			sys_put_le32(min[i], tmp);
+		} else if (item == 1) {
+			sys_put_le32(max[i], tmp);
+		} else if (item == 2) {
+			if (res) {
+				sys_put_le32(res[i], tmp);
+			} else {
+				memset(tmp, 0, 4);
+			}
+		}
+		net_buf_add_mem(buf, tmp, to_add);
+		length -= to_add;
+
+		if (++item == 3) {
+			item = 0;
+			i++;
+		}
+	}
+}
+
+static int get_clock_source_request(struct usbd_class_node *const node,
+				    const struct usb_setup_packet *const setup,
+				    struct net_buf *const buf)
+{
+	const uint32_t *frequencies;
+	int count;
+
+	/* Channel Number must be zero */
+	if (CONTROL_CHANNEL_NUMBER(setup) != 0) {
+		LOG_DBG("Clock source control with channel %d",
+			CONTROL_CHANNEL_NUMBER(setup));
+		return -EINVAL;
+	}
+
+	count = clock_frequencies(node, CONTROL_ENTITY_ID(setup), &frequencies);
+
+	if (CONTROL_SELECTOR(setup) == CS_SAM_FREQ_CONTROL) {
+		if (CONTROL_ATTRIBUTE(setup) == CUR) {
+			if (count == 1) {
+				layout3_cur_response(buf, setup->wLength,
+						     frequencies[0]);
+				return 0;
+			}
+			/* TODO: If there is more than one frequency supported,
+			 * call registered application API to determine active
+			 * sample rate.
+			 */
+		} else if (CONTROL_ATTRIBUTE(setup) == RANGE) {
+			layout3_range_response(buf, setup->wLength, frequencies,
+					       frequencies, NULL, count);
+			return 0;
+		}
+	} else {
+		LOG_DBG("Unhandled clock control selector 0x%02x",
+			CONTROL_SELECTOR(setup));
+	}
+
+	return -ENOTSUP;
+}
+
+static int uac2_control_to_host(struct usbd_class_node *const node,
+				const struct usb_setup_packet *const setup,
+				struct net_buf *const buf)
+{
+	entity_type_t entity_type;
+
+	if ((CONTROL_ATTRIBUTE(setup) != CUR) &&
+	    (CONTROL_ATTRIBUTE(setup) != RANGE)) {
+		return -ENOTSUP;
+	}
+
+	if (setup->bmRequestType == GET_CLASS_REQUEST_TYPE) {
```

Comment:
This is probably okay, but you only need to differentiate the request type class to the recipient interface and endpoint, otherwise the stack would not call this callback.

Reply:
Yes, Audio class does support interface and endpoint requests. Here, currently the absolute minimum of control requests are handled to satisfy the most basic example.